src/passes/OptimizeInstructions.cpp

```diff
@@ -1372,39 +1372,63 @@ struct OptimizeInstructions
-    if (curr->left->is<RefNull>()) {
-      std::swap(curr->left, curr->right);
-    }
-
-    // RefEq of a value to Null can be replaced with RefIsNull.
-    if (curr->right->is<RefNull>()) {
-      replaceCurrent(Builder(*getModule()).makeRefIs(RefIsNull, curr->left));
-    }
-  }
+  // Note on removing casts (which the following utilities, skipNonNullCast and
+  // skipCast do): removing a cast is potentially dangerous, as it removes
+  // information from the IR. For example:
+  //
+  //  (ref.is_func
+  //    (ref.as_func
+  //      (local.get $anyref)))
+  //
+  // The local has no useful type info here (it is anyref). The cast forces it
+  // to be a function, so we know that if we do not trap then the ref.is will
+  // definitely be 1. But if we removed the ref.as first (which we can do in
+  // traps-never-happen mode) then we'd not have the type info we need to
+  // optimize that way.
+  //
+  // To avoid such risks we should keep in mind the following:
+  //
+  //  * Before removing a cast we should use its type information in the best
+  //    way we can. Only after doing so should a cast be removed. In the exmaple
+  //    above, that means first seeing that the ref.is must return 1, and only
+  //    then possibly removing the ref.as.
+  //  * Do not remove a cast if removing it might remove useful information for
+  //    others. For example,
+  //
+  //      (ref.cast $A
+  //        (ref.as_non_null ..))
+  //
+  //    If we remove the inner cast then the outer cast becomes nullable. That
+  //    means we'd be throwing away useful information, which we should not do,
+  //    even in traps-never-happen mode and even if the wasm would validate
+  //    without the cast. Only if we saw that the parents of the outer cast
+  //    cannot benefit from non-nullability should we remove it.
+  //    Another example:
+  //
+  //      (struct.get $A 0
+  //        (ref.cast $B ..))
+  //
+  //    The cast only changes the type of the reference, which is consumed in
+  //    this expression and so we don't have more parents to consider. But it is
+  //    risky to remove this cast, since e.g. GUFA does benefit from such info:
+  //    it tells GUFA that we are reading from a $B here, and not the supertype
+  //    $A. If $B may contain fewer values in field 0 than $A, then GUFA might
+  //    be able to optimize better with this cast. But in traps-never-happen
+  //    mode we can assume that only $B can arrive here, which means GUFA should
+  //    be able to infer that even without the cast. Thus, in traps-never-happen
+  //    mode we should optimize out such a cast (assuming the code still
+  //    validates, of course), but not otherwise. (This may not always be the
+  //    optimal decision, say if GUFA hits a limitation and does not manage to
+  //    make the inference that only $B can arrive, but as it is important to
+  //    remove such redundant casts, we have to draw a line somewhere.)
```

Comment:
Has this implemented?

Reply:
In GUFA? Yes, to some extent. It may see that `$B` is the only thing that reaches a location, say if it comes from a `struct.new $B`. But it will hit limitations if more than one type is possible, etc., at least for now.

Reply:
It seems like it would probably be worth introducing a notion of "late" OptimizeInstructions in which we aggressively remove as many casts as possible without worrying about preserving information for follow-on optimizations.

Reply:
Yes, a "late" pass may be useful. The risk is the module becomes less optimizable later. But I can't see a way around that...